src/linker/Linker.Steps/IMarkAssemblyStep.cs

```diff
@@ -0,0 +1,33 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Mono.Cecil;
+
+namespace Mono.Linker.Steps
+{
+
+	/// <summary>
+	/// Extensibility point for custom steps that run during MarkStep,
+	/// for every marked assembly.
+	/// </summary>
+	public interface IMarkAssemblyStep
```

Comment:
I'm not sure this is the right step forward. You are adding ondemand step in the way it's not very clear to me how it should be used. I'm also not sure OnAssemblyMarked is actually the point where the custom steps would like to hook up. I'd expect them to ignore assembly-mark completely and hook up into OnTypeMarked or OnMethodMarked instead.

Secondly, does this also mean there is no way to actually mark assembly when it's loaded via custom step?

Reply:
My current thinking:
* We want the ability for custom steps to react to marking of certain things, for starters assemblies, but eventually also types and maybe even methods.
* Each item type will need its pipeline at least internally - we simply need a list of custom steps to run for each item type - for example for each assembly. Whether we explicitly expose the notion of pipelines or not is currently a secondary concern (I think we might as well)
* To not hurt perf, if a custom step wants to react to only assemblies, it should not need to in any way subscribe to type/methods. `SubStepDispatcher` solves this by making all sub-steps implement everything, but then it uses an enum to tell the system which of these the step is interested in. I would choose a more declarative approach. Have an interface for each, so (names obviously pending discussion):
  * `IMarkAssemblyStep`
  * `IMarkTypeStep`
  * `IMarkMethodStep`
* If we want to we could also create base interface `IMarkStep` which would have the `Initialize` method and all of the above would "inherit" from it.
* Each interface would have basically one method, something like `OnAssemblyMarked`/`OnTypeMarked`/`OnMethodMarked`. The semantics is rather obvious.
* Note that there's a difference between loading and assembly and marking an assembly - steps are allowed to load assemblies as they please (they can also freely call `.Resolve()`) and they are even allowed to mark additional assemblies. Once assembly is marked and the `MarkStep` gets to it, it would call all custom steps implementing `IMarkAssemblyStep` and "notify" them.
* As discussed in other PRs, marking in custom steps would be always deferred (using `Annotations` and actual marking happens once the `MarkStep` gets to it).
* I think we're doing `OnMarkAssembly` first because we think we'll need it to implement some of existing steps in the linker. Once we start porting over Xamarin steps we should definitely consider adding `OnMarkType` or maybe even `OnMarkMethod` (or maybe `OnMarkMember` ? not sure) - I think assembly and type should suffice for .NET 6, but we'll see.

Other considerations:
* Existing `IStep` infra would remain basically intact. The main difference is that it won't automatically get the entire assembly closure. It could load the closure on demand, but it won't ever be guaranteed to be complete (that's why we're doing #1666) if the step us running before `MarkStep`.
* Even the new custom steps could in theory load the closure aggressively, but it's not recommended for perf reasons. That said, linker won't react to assembly loading as an event with anything. The only event which has meaning is marking an assembly.
* "marking an assembly" in the above means that linker will preserve the assembly in some form in the output (trimmer or not). We will need to make sure that the `OnMarkAssembly` gets called for all assemblies written to the output (I don't know if today we internally actually mark `copy` assemblies for example, if not we should fix that).
* There will be certain overlap between the "preserve" mechanics in annotations (`SetPreserve` and `AddPreservedMethod`) and the `OnMarkType`. It will be possible to implement the preserve semantics with `OnMarkType` and simple marking. I don't think this is inherently bad, just two options which will end up doing the same thing.
* We could allow one step to potentially implement more than one such interface - they would all share the same instance (we would ever only create one instance of the custom step), but it could belong to multiple pipelines (this gets a bit tricky with ordering though, not sure yet)

Reply:
I pretty much agree with Vitek's thoughts - I see us moving in a direction where custom logic will just extend the marking logic by adding conditional dependencies like "source -> destination" which means "if source is marked, destination should be marked". The current mechanism for this is `AddPreservedMethod`, which requires custom steps to discover the "source" types by scanning for them eagerly, but in the future extension points like `IMarkTypeStep`/`OnMarkType` would be better since they avoid an extra scan of unmarked types.

`IMarkAssemblyStep` (or whatever we name it) would let most existing steps work with little to no modification, even when we add lazy loading, as long as they only use `AddPreservedMethod`/`SetPreserve`. The scope of the scanning will be reduced to marked assemblies only.

Any steps which need to scan for types to `Mark` unconditionally (rather than conditionally `AddPreservedMethod`/`SetPreserve`) will have two options:
- If it is enough to discover types in marked assemblies, they can use `IMarkTypeStep` without further modification
- If they need to discover types in otherwise unmarked assemblies, they can continue using `IStep` and look at the loaded closure (or explicitly load it, when we turn on lazy loading).

I don't have a strong preference on whether the hooks should be declared using interfaces or an enum similar to `SubStepTargets`. Either way, if we allow one custom step to hook into multiple events, we need to decide how to handle the filtering capability of `ISubStep`. The `IsActiveFor` check is nice because it currently prevents unnecessary calls to the substep for types/members/etc for assemblies where it's not active. With `IMarkAssemblyStep` it just works the same as it does today. But if we add, say `IMarkMethod`, then we would want to avoid calling the custom step method logic for a marked method inside of an assembly for which it's not active. Essentially some of the SubStepDispatcher logic would become built-in to MarkStep or the linker pipeline. I think just adding `IMarkAssemblyStep` is a good way to make progress, while leaving that possibility open.

Reply:
I think there are two needs which might be better to handle separately.

1. Run additional processing after assembly load

It's something that is needed primarily by linker steps. I'm not convinced we need to expose this functionality publicly at this point to custom steps therefore this can be implemented as simple
method calls. The important part is that this logic can be triggered from any step.
Something like this might be enough.

```c#
class AssemblyResolver
{
    public override AssemblyDefinition Resolve (AssemblyNameReference name, ReaderParameters parameters)
    {
        // some existing code which loads assembly

        // This could be even refactored into single class, etc.
        BlacklistStep.Process (context, assembly);
        LinkAttributesStep.Process (context, assembly);
    }
}
```

2. Conditional Marking

Many of the existing custom steps include logic which marks additional code based on a specific
pattern. This API should be exposed but I think we need to validate it first by trying to
rewrite a few of the existing custom steps. I'd like to see API along following the quick draft.

```c#
public interface IMarkStepProcessing
{
    void Initialize (LinkContext context, MarkProcessingRegistration register);
}

public class MarkProcessingRegistration
{
    public void AsemblyProcessed (Action<AssemblyDefinition> action);
    public void TypeProcessed (Action<TypeDefinition> action);
}
```

Reply:
I agree we should try to convert the steps which linker does always into non-step code - just like we've converted const prop and basically all of XML loading. Maybe there will be something which would still make sense as a "step" - we'll see as we go through these.

I also agree that it sounds like a good idea to try to convert a couple of Xamarin steps and design the new extensibility along the way.

1. I would like to avoid linker doing almost anything on assembly load. I think the current idea is that linker would do nothing eagerly. LinkAttributes will be loaded on demand when somebody asks for linker attributes on something from an assembly which didn't load them yet. Substitutions will be loaded also on demand once the constant prop touches something from a new assembly. Descriptors should only be processed when the assembly is marked so it will be called from the MarkAssembly (or nearby). There might be some tweaks to this with regard to copy assemblies and their dependencies, but hopefully not.

2. This looks good - I'm not sure I like the word "processed", but that's naming - and I suck at it :wink: